contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+// API Server will fail validation of all update requests that attempt to mutate 
+// the Data field. AnyStates may, however, be deleted.
+type AnyState struct {
+    metav1.TypeMeta
+    // +optional
+    metav1.ObjectMeta
+    // Data is a map of keys to the serialized objects that comprise the state 
+    // snapshot.
+    // Each key must consist of alphanumeric characters, '-', '_' or '.'.
+    // +optional
+    Data map[string]string
+}
+```
+
+## API Server
+The API Server must support the creation and deletion of AnyState objects.
+As we have no mechanism for declarative immutability, the API server must fail 
+any update request that updates the `.Data` field of an AnyState Object.
+
+## Controllers 
+This section is presented as a generalization of how an arbitrary controller 
+can use AnyState to persist a history of revisions to its specification type 
+Objects. The technique is applicable, without loss of generality, to the 
+existing Kuberentes controllers that have Pod as a generated type.
+
+When a controller detects a revision to the target Object state of a 
+specification type Object it will do the following.
+
+1. The controller will [create a snapshot](#version-snapshot-creation) of the 
+current target Object state.
+1. The controller will [reconstruct the history](#history-reconstruction) of 
+revisions to the Object's target Object state.
+1. The controller will test the current target Object state for 
+[equivalence](#version-equivalence) with all other versions in the Object's 
+revision history.
+    - If the current version is semantically equivalent to its immediate 
+    predecessor no update to the Object's target state has been performed.
+    - If the current version is equivalent to a version prior to its immediate 
+    predecessor, this indicates a rollback.
+    - If the current version is not equivalent to any prior version, this 
+    indicates an update or a roll forward.
+    - Controllers should use their status objects for book keeping with respect 
+    to current and prior revisions.
+1. The controller will 
+[reconcile its generated Objects](#target-object-state-reconciliation) 
+with the new target Object state.
+1. The controller will [maintain the length of its history](#history-maintenance) 
+to be less than the configured limit.
+
+### Version Snapshot Creation
+To take a snapshot of the target Object state contained in a specification type
+Object, a controller will do the following.
+
+1. The controller will serialize all the fields of the Object's target object 
+state using JSON encoding.
+1. The controller will populate the `.Data` of an AnyState object by 
+associating the serialized fields with keys that correspond to the fields' 
+names. 
+   - e.g `PodTemplateSpecs` should be stored at the `"Template"` key.
+1. The controller will store the specification type Objects `.Generation` at the 
+`"ControllerGeneration"` key.
```

Comment:
Array is:

```
Templates []History

type History struct {
  Purpose/Name string // i.e. "podTemplate"
  TypeMeta // the api and version of the historical snippet for use by the client
  Value runtime.RawExtension // the snippet
}
```

Purpose/Name would map back (same as the key in the map).

Another point - a pod template is serializable but a volume claim template is not (no type metadata).  So there has to be additional versioning data attached if you want to store things that are not top level objects .  You could alternatively require the controller to store a full object `PersistentVolumeClaim` instead of `VolumeClaimTemplate` and reconstitute the sub object from the full, or you could create a new `VolumeClaimTemplate` and attach object metadata (which is not used?).  Ideally we would only serialize exactly what we need, but we would need to specify a group and version for the data we specify.  

If serialized state can span two api groups/versions, we need to have distinct group versions.  If the state must be intrinsic, the controller revision may store a single type/version (and arbitrary snippets that it can understand for that version).  I kind of prefer the latter, which means you can serialize subsets of the object.

In proto serialization, do you want to store JSON?